### PR TITLE
Reading processing instructions

### PR DESCRIPTION
**Pull request by Dantemss**
- Added reading feature map fields to Content::Models::Book
- Added reading feature map to Ecosystem manifests
- Fragment splitter now uses the feature map
- Fixed HS iReading specs
- When no embed tag found, search for exercises by context tag
- Exercises with multiple LO's and/or CNXMOD tags in multiple pages now are imported/map into the highest numbered page (!!!)
- CNXMOD tags can now also be used to map exercises to pages across multiple book versions
- Added manifest format-related error messages
- Specs

### CODE REVIEW

File: app/subsystems/content/models/page.rb [hunk 1]
```diff
@@ -45,16 +45,32 @@ def ccs
...
+        fragments: fragment_splitter.split_into_fragments(snap_lab_node, 'snap-lab')
+      }
     end
   end
 
   protected
 
   def parser
-    OpenStax::Cnx::V1::Page.new(title: title, content: content)
+    @parser ||= OpenStax::Cnx::V1::Page.new(title: title, content: content)
```
Comment (jpslav) [inline app/subsystems/content/models/page.rb]: good catch

File: app/subsystems/tasks/assistants/extra_assignment_assistant.rb [hunk 2]
```diff
@@ -24,45 +26,39 @@ def initialize(task_plan:, taskees:)
...
-      pages: @ecosystem.pages_by_ids(page_ids),
+      pages: ecosystem.pages_by_ids(page_ids),
       page_id_to_snap_lab_id: page_to_snap_lab
     }
   end
 
   def build_extra_task(pages:, page_id_to_snap_lab_id:)
     task = build_task
 
+    reset_used_exercises
```
Comment (jpslav) [inline app/subsystems/tasks/assistants/extra_assignment_assistant.rb]: I'm a little worried that some day down the line we're going to forget to call this reset method in some later assistant... but maybe I'm worried over nothing.

File: app/subsystems/tasks/assistants/i_reading_assistant.rb [hunk 3]
```diff
@@ -65,60 +59,18 @@ def build_ireading_task(pages:, taskee:)
...
       task_type: :reading,
       title:     title,
       description: description,
       feedback_at: Time.now
     ]
-    AddSpyInfo[to: task, from: @ecosystem]
-    return task
-  end
-
-  def task_fragments(task:, fragments:, fragment_title:, page:, related_content:)
```
Comment (jpslav) [inline app/subsystems/tasks/assistants/i_reading_assistant.rb]: super cool to pull this code up into `FragmentAssistant`!

File: lib/openstax/cnx/v1/fragment_splitter.rb [hunk 4]
```diff
@@ -1,98 +1,116 @@
...
+    def split_into_fragments(root, type = nil)
+      result = [root.dup]
+      type_string = type.to_s
+
+      processing_instructions.each do |processing_instruction|
+        next if processing_instruction.css.blank? ||
+                processing_instruction.fragments.nil? ||
+                (
+                  (!processing_instruction.only.nil? &&
+                   ![processing_instruction.only].flatten.map(&:to_s).include?(type_string)) ||
```
Comment (jpslav) [inline lib/openstax/cnx/v1/fragment_splitter.rb]: `flat_map`?

File: lib/openstax/cnx/v1/fragment_splitter.rb [hunk 5]
```diff
@@ -1,98 +1,116 @@
...
+                  (!processing_instruction.except.nil? &&
+                   [processing_instruction.except].flatten.map(&:to_s).include?(type_string))
+                )
```
Comment (jpslav) [inline lib/openstax/cnx/v1/fragment_splitter.rb]: Why skip the instruction if the `only` matches the type string?  (I understand skipping if `except` matches)

Also I think the nesting can be simplified here.  This has `A || B || (C || D)` which can be `A || B || C || D`.

File: spec/factories/content/books.rb [hunk 6]
```diff
@@ -9,6 +9,22 @@
     content { contents.to_json }
     uuid { SecureRandom.uuid }
     version { Random.rand(1..10) }
+    reading_processing_instructions {
+      [
+        { css: '.ost-reading-discard, .os-teacher, [data-type="glossary"]',
+          fragments: [], except: 'snap-lab' },
```
Comment (jpslav) [inline spec/factories/content/books.rb]: What does "excepting" `snap-lab` do here since `fragments` is `[]`?

**Comment by jpslav** [comment]
Looking good!

File: lib/openstax/cnx/v1/fragment_splitter.rb (hunk 5 above)
Comment (Dantemss) [inline lib/openstax/cnx/v1/fragment_splitter.rb]: There's a `!` on that line :D and ok I'll simplify the nesting.

File: lib/openstax/cnx/v1/fragment_splitter.rb (hunk 4 above)
Comment (Dantemss) [inline lib/openstax/cnx/v1/fragment_splitter.rb]: `flat_map` is equivalent to `flatten(1).map` but I really want to `flatten` everything here so I don't think it applies...

File: spec/factories/content/books.rb (hunk 6 above)
Comment (Dantemss) [inline spec/factories/content/books.rb]: Each instruction removes the matching nodes unless `"node"` is in the `fragments` list. So `fragments: []` is the discard instruction. Skipping it for `snap-lab` prevents the discard instruction from applying to snap-labs.

**Comment by Dantemss** [comment]
I decided to split the try another after reading step stuff into a separate PR so this one can be finished.

**Comment by Dantemss** [comment]
@jpslav feel free to review this whenever. It's only missing specs.

**Comment by Dantemss** [comment]
OK I think I am happy with the number of tests for this now. Merge away when ready.

**Pull request merged**